src/engine/Map.cpp

```diff
@@ -0,0 +1,530 @@
+
+        {"demo", NULL, 0},
+        {"demo2", NULL, 0},
+        {"easy", NULL, 0},
+
+        {"genbrack", NULL, MAP_MODE_GENERATED},
+        {"genchalk", NULL, MAP_MODE_GENERATED},
+        {"gendesrt", NULL, MAP_MODE_GENERATED},
+        {"genlava", NULL, MAP_MODE_GENERATED},
+        {"genmud", NULL, MAP_MODE_GENERATED},
+        {"gensnow", NULL, MAP_MODE_GENERATED},
+
+        {"hillbase", NULL, 0},
+
+        {"airaid", NULL, 0},
+        {"canal", NULL, 0},
+        {"climb", NULL, 0},
+        {"faceoff", NULL, 0},
+        {"hell1", NULL, 0},
+        {"hiroad", NULL, 0},
+        {"hwd", NULL, 0},
+        {"newsnu", NULL, 0},
+        {"pdrag", NULL, 0},
+        {"quack", NULL, 0},
+        {"tdd", NULL, 0},
+};
+#endif
+
+Map::Map(const GameModeSetup &mode) {
+    LogDebug("Loading map, %s, in mode %u\n", mode.map, mode.game_mode);
+    /* map manager now worries about resetting map state etc. */
+
+    MapManifest *desc = MapManager::GetInstance()->GetManifest(mode.map);
+    if(desc == nullptr) {
+        LogWarn("Failed to get map descriptor, %s!\n", mode.map);
+    } else {
+        LogDebug("Found map descriptor!\n %s\n %s", desc->name.c_str(), desc->description.c_str());
+        if(mode.game_mode != MAP_MODE_EDITOR && !(desc->flags & mode.game_mode)) {
+            throw std::runtime_error("This mode is unsupported by this map, " + std::string(mode.map) + ", aborting!\n");
+        }
+
+        name_           = desc->name;
+        description_    = desc->description;
+        sky_            = desc->sky;
+    }
+
+    std::string base_path = "maps/" + std::string(mode.map) + "/";
+    std::string p = u_find(std::string(base_path + std::string(mode.map) + ".pmg").c_str());
+    if(!plFileExists(p.c_str())) {
+        throw std::runtime_error("PMG, " + p + ", doesn't exist!\n");
+    }
+
+    LoadTiles(p);
+
+    p = u_find(std::string(base_path + std::string(mode.map) + ".pog").c_str());
+    if (!plFileExists(p.c_str())) {
+        throw std::runtime_error("POG, " + p + ", doesn't exist!\n");
+    }
+
+    LoadSpawns(p);
+    LoadTextures(p);
+
+    sky_model_ = LoadModel("skys/skydome", true);
+
+    GenerateOverview();
+
+    Reset();
+}
+
+Map::~Map() {
+    if(!tile_textures_.empty()) {
+        LogDebug("Freeing %u textures...\n", tile_textures_.size());
+        for(auto texture : tile_textures_) {
+            plDeleteTexture(texture, true);
+        }
+    }
+
+    if(overview_ != nullptr) {
+        plDeleteTexture(overview_, true);
+    }
+
+    for (auto &sky_texture : sky_textures_) {
+        if(sky_texture == nullptr) {
+            break;
+        }
+
+        plDeleteTexture(sky_texture, true);
+    }
+
+    if(sky_model_ != nullptr) {
+        plDeleteModel(sky_model_);
+    }
+}
+
+MapChunk *Map::GetChunk(const PLVector2 &pos) {
+    uint idx = ((uint)(pos.x) / MAP_CHUNK_PIXEL_WIDTH) + (((uint)(pos.y) / MAP_CHUNK_PIXEL_WIDTH) * MAP_CHUNK_ROW);
+    if(idx >= chunks_.size()) {
```

Comment:
Is the width of the map known at this point? Probably worth expanding the check to see if x is past the end of the row, else x going out of bounds will silently wrap into the next row.